### PR TITLE
Implement methods for CSV and Ndjson batch adds

### LINKED ISSUE #238: NDJSON/CSV methods to add and update documents
⚠️ This issue is generated, it means the nameing might be done differently in this package (ex: `add_documents_json` instead of `addDocumentsJson`). Keep the already existing way of naming in this package to stay idiomatic with the language and this repository.

📣 We strongly recommend doing multiple PRs to solve all the points of this issue

[MeiliSearch v0.23.0](https://github.com/meilisearch/MeiliSearch/releases/tag/v0.23.1) introduces two changes:
- new valid formats to push data files, additionally to the JSON format: CSV and NDJSON formats.
- it enforces the `Content-type` header for every route requiring a payload (`POST` and `PUT` routes)

---

Here are the expected changes to completely close the issue:

- [x] Currently, the SDKs always send `Content-Type: application/json` to every request. Only the `POST` and `PUT` requests should send the `Content-Type: application/json` and not the `DELETE` and `GET` ones.

- [x] Add the following methods and 🔥 **the associated tests** 🔥 to ADD the documents. Depending on the format type (`csv` or `ndjson`) the SDK should send `Content-Type: application/x-dnjson` or `Content-Type: text/csv`)
  - [x] `addDocumentsJson(string docs, string primaryKey)`
  - [x] `addDocumentsCsv(string docs, string primaryKey)`
  - [x] `addDocumentsCsvInBatches(string docs, int batchSize, string primaryKey)`
  - [x] `addDocumentsNdjson(string docs, string primaryKey)`
  - [x] `addDocumentsNdjsonInBatches(string docs, int batchSize, string primaryKey)`

- [x] Add the following methods and 🔥 **the associated tests** 🔥 to UPDATE the documents. Depending on the format type (`csv` or `ndjson`) the SDK should send `Content-Type: application/x-dnjson` or `Content-Type: text/csv`)
  - [x] `updateDocumentsJson(string docs, string primaryKey)`
  - [x] `updateDocumentsCsv(string docs, string primaryKey)`
  - [x] `updateDocumentsCsvInBatches(string docs, int batchSize, string primaryKey)`
  - [x] `updateDocumentsNdjson(string docs, string primaryKey)`
  - [x] `updateDocumentsNdjsonInBatches(string docs, int batchSize, string primaryKey)`

`docs` are the documents sent as `String`
`primaryKey` is [the primary key of the index](https://docs.meilisearch.com/learn/core_concepts/indexes.html#primary-key)
`batchSize` is the size of the batch. Example: you can send 2000 documents in raw String in `docs` and ask for a `batchSize` of 1000, so your documents will be sent to MeiliSearch in two batches.

Example of PRs:
- in PHP SDK: https://github.com/meilisearch/meilisearch-php/pull/235
- in Python SDK: https://github.com/meilisearch/meilisearch-python/pull/329

----------

Related to: https://github.com/meilisearch/integration-guides/issues/146

If this issue is partially/completely implemented, feel free to let us know.

### PR DESCRIPTION
**Pull request by karlomikus**
# Pull Request

## Related issue
Fixes #238

## What does this PR do?
- Implements methods for batch adding of documents in csv and ndjson format

## PR checklist
Please check if your PR fulfills the following requirements:
- [x] Does this PR fix an existing issue, or have you listed the changes applied in the PR description (and why they are needed)?
- [x] Have you read the contributing guidelines?
- [x] Have you made sure that the title is accurate and descriptive of the changes?

## Notes
This is using `preg_split` and `implode` to chunk strings by newline. Let me know if there is a better way to do this.

If this is fine I can implement the rest of the methods from the issue.

### CODE REVIEW

File: src/Endpoints/Delegates/HandlesDocuments.php [hunk 1]
```diff
@@ -100,6 +120,31 @@ private function assertValidDocumentId($documentId): void
...
+
+    private static function batchNdjsonString(string $documents, int $batchSize): Generator
+    {
+        $documents = preg_split("/\r\n|\n|\r/", trim($documents));
+
+        $batches = array_chunk($documents, $batchSize);
+        foreach ($batches as $batch) {
+            $batch = implode("\n", $batch);
+            yield $batch;
+        }
```
Comment (brunoocasali) [inline src/Endpoints/Delegates/HandlesDocuments.php:145]:
```suggestion
        $documents = preg_split("/\r\n|\n|\r/", trim($documents));
        $batches = array_chunk($documents, $batchSize);
        
        foreach ($batches as $batch) {
            $batch = implode("\n", $batch);

            yield $batch;
        }
```

File: src/Endpoints/Delegates/HandlesDocuments.php [hunk 2]
```diff
@@ -100,6 +120,31 @@ private function assertValidDocumentId($documentId): void
...
+        $documents = preg_split("/\r\n|\n|\r/", trim($documents));
+        $csvHeader = $documents[0];
+        array_shift($documents);
+
+        $batches = array_chunk($documents, $batchSize);
+        foreach ($batches as $batch) {
+            array_unshift($batch, $csvHeader);
+            $batch = implode("\n", $batch);
+            yield $batch;
+        }
```
Comment (brunoocasali) [inline src/Endpoints/Delegates/HandlesDocuments.php:134]:
```suggestion
        array_shift($documents);
        $batches = array_chunk($documents, $batchSize);

        foreach ($batches as $batch) {
            array_unshift($batch, $csvHeader);
            $batch = implode("\n", $batch);

            yield $batch;
        }
```

**Review by brunoocasali — COMMENTED** [review]
Hi @karlomikus, I'm so sorry for the delay in this review!

But I'm very satisfied with how you submitted this PR. It is very good for me! You can submit the next ones you mentioned in the notes section ;)

I just left two suggestions regarding the structure of the code. I think the code will be even easier to read with the suggestions.

Thanks a lot for contributing to Meilisearch! ❤️

**Review by brunoocasali — DISMISSED** [review]
Thank you so much! 🤘 

bors merge

**Comment by brunoocasali** [comment]
@karlomikus I tried to merge, but we still have some issues to fix regarding the phpstan can you check that? :)

**Comment by brunoocasali** [comment]
bors merge

**Comment by karlomikus** [comment]
I've fixed phpstan errors, I completely missed them :)

I think contributing guidelines should be updated to include `composer phpstan` run command.

**Review by brunoocasali — APPROVED** [review]
Awesome, thanks a lot for your contribution @karlomikus 🤘 

bors merge

**Comment by brunoocasali** [comment]
_This message is sent automatically_

Thanks again for contributing to Meilisearch :heart:
If you are participating in [Hacktoberfest](https://blog.meilisearch.com/contribute-hacktoberfest-2022/), and you would like to receive some gift from Meilisearch too, please complete [this form](https://meilisearch.typeform.com/to/idt5qugY).

**Pull request merged**